.changeset/neat-falcons-add.md

```diff
@@ -0,0 +1,63 @@
+---
+'@shopify/remix-oxygen': patch
+'@shopify/hydrogen': patch
+'@shopify/cli-hydrogen': patch
+'@shopify/create-hydrogen': patch
+---
+
+Subrequest Profiler (stable) - Provides an overview of network requests happening on the server side
+
+#### How to use:
+
+1. Run `h2 dev`
+2. Visit http://localhost:3000/subrequest-profiler
+
+#### Set request display name with `storefront.query`:
+
+```tsx
+context.storefront.query(
+  HOMEPAGE_FEATURED_PRODUCTS_QUERY,
+  {
+    displayName: 'Feature products',
+    variables: {
+      country,
+      language,
+    },
+  },
+)
+```
+
+#### Set request debug information with `createWithCache`:
+
+```tsx
+const withCache = createWithCache({
+  cache,
+  waitUntil,
+  request,
+});
+
+const catFact = async () => {
+  return await withCache<CatFact>(
+    ['Some unique cache keys'],
+    CacheLong(),
+    ({addDebugData}) => {
+      return fetch('https://some-3p-endpoint.com').then(async (res) => {
+        if (process.env.NODE_ENV === 'development') {
+          addDebugData({
+            displayName: '3p endpoint display name',
+            url: 'https://some-3p-endpoint.com',
+            responseInit: {
+              status: res.status,
+              statusText: res.statusText,
+              headers: Array.from(res.headers.entries()),
+            }
+          });
```

Comment:
IMO this is could be much simpler. Why not just `addDebugData({displayName: 'something', response: res})`

Reply:
good point

Reply:
Maybe there are multiple responses, though?

Reply:
There can be multiple responses but it's up to the dev to supply us what they want to show on the debug. They can create a new Response with all the data they want from their responses.

The chart timing for `createWithCache` will always the total duration of all the responses within the `actionFn`. Dev could also wrap `withCache` with another `withCache`. Network waterfall chart may look a bit funny but it kinda works.